src/java/org/apache/cassandra/db/Directories.java

```diff
@@ -274,6 +294,76 @@ public boolean accept(File file)
+    }
+
+    @VisibleForTesting
+    static Runnable getVerifyDiskHasEnoughUsableSpaceRunnable()
+    {
+        return () -> {
+            try {
+                Directories.verifyDiskHasEnoughUsableSpace();
+            } catch (FSWriteError e) {
+                // If node is already disabled propagating the exception from scheduled executor is redundant and clogs logs.
+                if (!StorageService.instance.isNodeDisabled()) {
+                    throw e;
+                }
+            }
+        };
+    }
+
+    public static void verifyDiskHasEnoughUsableSpace()
+    {
+        Map<DataDirectory, Double> dirsToUtilization = Arrays.stream(dataDirectories)
+                                                             .collect(Collectors.toMap(
+                                                             Function.identity(),
+                                                             dir -> {
+                                                                 long total = dir.getTotalSpace();
+                                                                 long used = total - dir.getAvailableSpace();
+                                                                 return (double) used / total;
+                                                             })
+                                                             );
+
+        Map.Entry<DataDirectory, Double> maxPathToUtilization = dirsToUtilization
+                                                                .entrySet()
+                                                                .stream()
+                                                                .max(Map.Entry.comparingByValue())
+                                                                .orElseThrow(() -> new RuntimeException(
+                                                                "Failed to filter most full data directory"));
+        logger.debug(String.format("Highest data directory disk utilization on path %s (%f)",
+                                   maxPathToUtilization.getKey().location, maxPathToUtilization.getValue()));
+        double threshold = DatabaseDescriptor.getMaxDiskUtilizationThreshold();
+        double currUtilization = maxPathToUtilization.getValue();
+        if (currUtilization >= threshold)
+        {
+            ExceededDiskThresholdException cause = new ExceededDiskThresholdException(
+                                    maxPathToUtilization.getKey().location, currUtilization, threshold);
+            throw new FSWriteError(cause, maxPathToUtilization.getKey().location);
+        }
+
+        if (currUtilization < threshold && StorageService.instance.isNodeDisabled() && StorageService.instance.isSetupCompleted())
+        {
+            Set<Map<String, String>> nonTransientErrors = StorageService.instance.getNonTransientErrors();
+            boolean isOnlyExceededDisk = true;
+            for (Map<String, String> nonTransientError : nonTransientErrors)
```

Comment:
see my comment below, but if we make exceeded disk a transient error, then this block simplifies to `if (nonTransientErrors.isEmpty())`